src/slurm_plugin/cluster_event_publisher.py

```diff
@@ -0,0 +1,408 @@
+# Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License").
+# You may not use this file except in compliance with the
+# License. A copy of the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "LICENSE.txt" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+# OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import logging
+import sys
+import traceback
+from collections import ChainMap
+from datetime import datetime, timezone
+from typing import Callable, Dict, Iterable, List
+
+from slurm_plugin.common import log_exception
+from slurm_plugin.slurm_resources import DynamicNode, SlurmNode, StaticNode
+
+logger = logging.getLogger(__name__)
+
+_LAUNCH_FAILURE_GROUPING = {
+    **{failure: "ice-failures" for failure in [*SlurmNode.EC2_ICE_ERROR_CODES, "LimitedInstanceCapacity"]},
+    **{failure: "vcpu-limit-failures" for failure in ["VcpuLimitExceeded"]},
+    **{failure: "volume-limit-failures" for failure in ["VolumeLimitExceeded", "InsufficientVolumeCapacity"]},
+    **{failure: "custom-ami-errors" for failure in ["InvalidBlockDeviceMapping"]},
+    **{failure: "iam-policy-errors" for failure in ["UnauthorizedOperation", "AccessDeniedException"]},
+}
+
+
+class ClusterEventPublisher:
+    """Class for generating structured log events for cluster."""
+
+    def __init__(self, event_publisher: Callable = lambda *args, **kwargs: None, max_list_size=100):
+        self._publish_event = event_publisher
+        self._max_list_size = max_list_size
+
+    @property
+    def publish_event(self):
+        return self._publish_event
+
+    @staticmethod
+    def timestamp():
+        return datetime.now(timezone.utc).isoformat(timespec="milliseconds")
+
+    @staticmethod
+    def create_with_default_publisher(
+        event_logger, cluster_name, node_role, component, instance_id, max_list_size=100, **global_args
+    ):
+        """Create an instance of CreateEventPublisher with the standard event publisher."""
```

Comment:
“an instance of CreateEventPublisher”
do you mean "an instance of ClusterEventPublisher"?

Reply:
Thanks, fixed it.